File: app/batch_writer.py
"""Batched write path with a write-ahead log (WAL).

Every POST /search is a tiny write. Writing each one straight to SQLite would
mean one transaction per search — terrible write amplification under load.
Instead:

  1. record(query) appends the raw query to an append-only WAL file (crash
     durability) and bumps an in-memory aggregation buffer (query -> pending
     delta). The live trie count is bumped immediately so suggestions react
     instantly.
  2. A background flusher persists the *aggregated* buffer to SQLite in ONE
     transaction when the buffer hits a size threshold or a time interval
     elapses, then truncates the WAL.

Aggregation is the win: 10,000 searches for "iphone" become a single
`count = count + 10000` row write. We expose this as a write-reduction ratio.

Crash model: if the process dies before a flush, the WAL still holds every
recorded search; on the next startup recover() replays it into SQLite before the
trie is built, so no search is lost.

The flush holds the buffer lock for its whole duration (swap -> SQLite write ->
WAL truncate). That briefly serializes concurrent searches, but flushes are
infrequent (~1/s) and write a single transaction, so the cost is negligible and
it keeps the WAL/SQLite/buffer trio strictly consistent (no lost-on-truncate
race).
"""
from __future__ import annotations

import threading
import time
from collections import defaultdict
from pathlib import Path

from app.config import config
from app.db import Database
from app.trie import Trie


class BatchWriter:
    def __init__(self, db: Database, trie: Trie, trending=None, wal_path: Path | None = None):
        self.db = db
        self.trie = trie
        self.trending = trending  # Phase 5: record recency on each search
        self.wal_path = Path(wal_path or config.paths.wal)
        self.wal_path.parent.mkdir(parents=True, exist_ok=True)

        self._buffer: dict[str, int] = defaultdict(int)
        self._lock = threading.Lock()
        self._wal = open(self.wal_path, "a", buffering=1)  # line-buffered

        # ---- write-reduction stats ----
        self.logical_writes = 0   # individual searches recorded
        self.physical_writes = 0  # rows written to SQLite across all flushes
        self.flushes = 0
        self.last_flush_ts = time.time()
        self.recency_invalidations = 0  # Phase 5

        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    # ---- recovery (called at startup, before trie build) ----
    @staticmethod
    def recover(db: Database, wal_path: Path | None = None) -> int:
        """Replay an un-flushed WAL into SQLite. Returns #searches recovered."""
        wal_path = Path(wal_path or config.paths.wal)
        if not wal_path.exists() or wal_path.stat().st_size == 0:
            return 0
        agg: dict[str, int] = defaultdict(int)
        n = 0
        with wal_path.open() as f:
            for line in f:
                q = line.rstrip("\n")
                if q:
                    agg[q] += 1
                    n += 1
        if agg:
            db.increment_many(agg)
        wal_path.write_text("")  # truncate after durable persist
        return n

    # ---- hot path ----
    def record(self, query: str) -> None:
        q = " ".join((query or "").split()).lower()
        if not q:
            return
        with self._lock:
            self._wal.write(q + "\n")          # durability first
            self._buffer[q] += 1
            self.logical_writes += 1
            # live in-memory update so suggestions reflect the search immediately
            entry = self.trie.get(q)
            if entry is None:
                entry = self.trie.insert(q, config.initial_count)
            else:
                entry.count += 1
            if self.trending is not None:
                self.trending.record(entry)
            buffer_full = len(self._buffer) >= config.batch_max_buffer
        if buffer_full:
            self.flush(reason="size")

    # ---- flush ----
    def flush(self, reason: str = "interval") -> int:
        with self._lock:
            if not self._buffer:
                return 0
            snapshot = dict(self._buffer)
            self._buffer.clear()
            rows = self.db.increment_many(snapshot)      # one transaction
            self._wal.flush()
            self._wal.seek(0)
            self._wal.truncate()                          # WAL consumed
            self.physical_writes += rows
            self.flushes += 1
            self.last_flush_ts = time.time()
            if self.trending is not None:
                self.recency_invalidations += self.trending.on_flush(snapshot)
        return rows

    # ---- background flusher ----
    def start(self) -> None:
        if self._thread is not None:
            return
        self._thread = threading.Thread(target=self._loop, name="batch-flusher", daemon=True)
        self._thread.start()

    def _loop(self) -> None:
        interval = config.flush_interval_s
        while not self._stop.wait(interval):
            try:
                self.flush(reason="interval")
            except Exception as e:  # pragma: no cover - keep flusher alive
                print(f"[batch] flush error: {e}")

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=2)
        self.flush(reason="shutdown")
        self._wal.close()

    # ---- stats ----
    def stats(self) -> dict:
        with self._lock:
            logical = self.logical_writes
            physical = self.physical_writes
            pending = len(self._buffer)
        reduction = (1 - physical / logical) if logical else 0.0
        return {
            "logical_writes": logical,          # individual searches received
            "physical_writes": physical,        # rows written to SQLite
            "pending_in_buffer": pending,
            "flushes": self.flushes,
            "write_reduction": round(reduction, 4),
            "write_reduction_pct": round(reduction * 100, 2),
            "avg_rows_per_flush": round(physical / self.flushes, 1) if self.flushes else 0,
            "recency_invalidations": self.recency_invalidations,
        }

File: src/batchWriter.js
// Batch writer: collects search-count increments and flushes them to the
// primary store in aggregated batches instead of one DB write per search.
//
// Pipeline per search:
//   record(query)
//     -> append "ts<TAB>query" to the write-ahead log (durable, sequential)
//     -> buffer.set(query, delta + 1)               (in-memory aggregation)
//     -> trending.record(query)                     (recency counter)
//     -> flush early if the buffer holds >= maxBufferSize distinct queries
//
// flush() (also fired on a periodic timer):
//   1. swap out the buffer
//   2. build aggregated entries with persisted recency snapshots
//   3. db.applyBatch(entries)        -- ONE transaction for the whole batch
//   4. trie.increment(...)           -- keep the in-memory index live
//   5. cache.invalidatePrefixesOf(q) -- drop stale cached prefix results
//   6. truncate the WAL              -- those writes are now durable in the DB
//
// Write-reduction: N repeated searches for the same query become a single
// "+N" row-write; many distinct queries share one transaction.
//
// Crash trade-off: a process crash between flushes loses nothing that reached
// the WAL — on restart recover() replays the log into the DB. The residual
// window is power-loss between the OS buffering the append and it hitting disk
// (we don't fsync per line, to keep the write path cheap); for stronger
// guarantees, fsync each WAL append at the cost of latency. Without the WAL,
// a crash would lose all un-flushed in-memory increments.

import fs from 'node:fs';
import { config } from './config.js';
import { metrics } from './metrics.js';

export class BatchWriter {
  constructor({ db, trie, cache, trending, opts = {} }) {
    this.db = db;
    this.trie = trie;
    this.cache = cache;
    this.trending = trending;
    this.flushIntervalMs = opts.flushIntervalMs ?? config.batch.flushIntervalMs;
    this.maxBufferSize = opts.maxBufferSize ?? config.batch.maxBufferSize;
    this.walPath = opts.walPath ?? config.paths.wal;

    this.buffer = new Map(); // query -> aggregated delta
    this.timer = null;
    this.walFd = null;
    this.lastFlush = { entries: 0, searchesCollapsed: 0, at: 0 };
  }

  start() {
    // Open the WAL for appending and keep the fd for fast sequential writes.
    this.walFd = fs.openSync(this.walPath, 'a');
    this.timer = setInterval(() => this.flush(), this.flushIntervalMs);
    // Don't keep the event loop alive solely for the flush timer.
    if (this.timer.unref) this.timer.unref();
  }

  // Replay any WAL left over from a previous crash, before serving traffic.
  // Returns the number of recovered search events.
  recover(now = Date.now()) {
    if (!fs.existsSync(this.walPath)) return 0;
    const raw = fs.readFileSync(this.walPath, 'utf8');
    if (!raw.trim()) return 0;

    const agg = new Map(); // query -> { delta, lastTs }
    let recovered = 0;
    for (const line of raw.split('\n')) {
      if (!line) continue;
      const tab = line.indexOf('\t');
      if (tab === -1) continue;
      const ts = parseInt(line.slice(0, tab), 10);
      const query = line.slice(tab + 1);
      if (!query) continue;
      const cur = agg.get(query) || { delta: 0, lastTs: 0 };
      cur.delta += 1;
      cur.lastTs = Number.isFinite(ts) ? Math.max(cur.lastTs, ts) : now;
      agg.set(query, cur);
      // Rebuild recency from the logged timestamps.
      this.trending.record(query, Number.isFinite(ts) ? ts : now);
      recovered++;
    }

    // Apply recovered increments directly to DB + trie, then clear the WAL.
    const entries = [];
    for (const [query, { delta }] of agg) {
      const snap = this.trending.snapshotFor(query, now);
      entries.push({
        query,
        countDelta: delta,
        recentScore: snap.recentScore,
        recentTs: snap.recentTs,
      });
      this.trie.increment(query, delta);
    }
    if (entries.length) {
      this.db.applyBatch(entries, now);
      metrics.searchesReceived += recovered; // count them as received
    }
    fs.truncateSync(this.walPath, 0);
    return recovered;
  }

  // Record a single search submission.
  record(query, now = Date.now()) {
    metrics.searchesReceived++;
    metrics.api.search++;
    // 1) durable append (sequential write — cheap relative to a DB txn)
    if (this.walFd !== null) {
      fs.writeSync(this.walFd, `${now}\t${query}\n`);
    }
    // 2) in-memory aggregation
    this.buffer.set(query, (this.buffer.get(query) || 0) + 1);
    // 3) recency counter
    this.trending.record(query, now);
    // 4) size-based flush trigger
    if (this.buffer.size >= this.maxBufferSize) this.flush(now);
  }

  // Flush the aggregation buffer to the primary store as one batch.
  flush(now = Date.now()) {
    if (this.buffer.size === 0) return { entries: 0, searchesCollapsed: 0 };

    const pending = this.buffer;
    this.buffer = new Map(); // swap so new searches accumulate independently

    let searchesCollapsed = 0;
    const entries = [];
    for (const [query, delta] of pending) {
      searchesCollapsed += delta;
      const snap = this.trending.snapshotFor(query, now);
      entries.push({
        query,
        countDelta: delta,
        recentScore: snap.recentScore,
        recentTs: snap.recentTs,
      });
    }

    // 1) durable primary-store write (single transaction)
    this.db.applyBatch(entries, now);

    // 2) keep the in-memory Trie counts live (no DB read needed)
    // 3) invalidate cached prefix results that could now be stale
    for (const e of entries) {
      this.trie.increment(e.query, e.countDelta);
      this.cache.invalidatePrefixesOf(e.query);
    }

    // 4) the buffered writes are now durable in the DB — clear the WAL
    if (this.walFd !== null) fs.ftruncateSync(this.walFd, 0);

    this.lastFlush = { entries: entries.length, searchesCollapsed, at: now };
    return this.lastFlush;
  }

  stats() {
    return {
      bufferedDistinctQueries: this.buffer.size,
      flushIntervalMs: this.flushIntervalMs,
      maxBufferSize: this.maxBufferSize,
      lastFlush: this.lastFlush,
    };
  }

  // Flush remaining buffer and release resources (called on shutdown).
  shutdown() {
    if (this.timer) clearInterval(this.timer);
    this.flush();
    if (this.walFd !== null) {
      fs.closeSync(this.walFd);
      this.walFd = null;
    }
  }
}

export default BatchWriter;
